rfd/0234-in-band-mfa-ssh-sessions.md

```diff
@@ -0,0 +1,767 @@
+service](https://github.com/gravitational/Teleport.e/blob/master/rfd/0024e-access-control-decision-api.md). Upon
+receiving a permit from the Decision service, the Proxy will staple the permit to the connection and open a connection
+to the respective cluster's SSH service.
+
+The SSH service will authenticate the client using the provided client certificate and the stapled permit. The SSH
+service will then check if MFA is required for the session by examining the permit.
+
+If MFA is _not required_, the SSH service will then proceed to establish the SSH session.
+
+If MFA _is required_, the SSH service will send a Protobuf [`MFAPrompt`
+message](#ssh-keyboard-interactive-authentication) via the SSH [keyboard-interactive
+channel](https://www.rfc-editor.org/rfc/rfc4256) to inform the client that MFA is needed. The client must then call the
+`CreateChallenge` RPC on the MFA service, providing a _Session Identifying Payload (SIP)_ and any relevant metadata. The
+SIP is an SSH session hash computed from SSH session state and is used to bind the MFA challenge to the specific
+session. Both the SSH client and the SSH server can independently compute the SIP from session state.
+
+The MFA service will [store the Session Identifying Payload (SIP)](#storing-session-identifying-payloads) and respond
+with an MFA challenge for the client to solve.
+
+Next, the client solves the MFA challenge and calls `ValidateChallenge` with the MFA response. The MFA service validates
+the MFA response and stores the `ValidatedChallenge` to the local backend.
+
+In addition, if the target cluster is a leaf cluster, a watcher in the reverse tunnel server monitors
+`ValidateChallenge` events for its cluster. When a `ValidateChallenge` is created for a leaf cluster, the reverse tunnel
+server calls the `CreateValidatedChallenge` RPC using its auth client to forward the challenge to the leaf cluster's MFA
+service.
+
+This ensures that validated MFA challenges are available in the appropriate cluster for retrieval during SSH session
+establishment.
+
+If validation is successful, the MFA service responds to the client with a confirmation that the challenge has been
+validated. If validation fails, the MFA service responds with an `Access Denied: Invalid MFA response` error.
+
+The client then sends a [`MFAPromptResponse`](#ssh-keyboard-interactive-authentication) message to the SSH service,
+instructing it to retrieve the validated challenge from the MFA service.
+
+The SSH service retrieves the validated challenge from the MFA service using the `GetValidatedChallenge` RPC. The SSH
+service independently computes the SIP from session state and verifies it matches the SIP retrieved from the MFA service
+in the validated challenge.
+
+If validation succeeds, the SSH session is established. If not, access is denied with an `Access Denied: Invalid MFA
+response` error. If the client does not complete MFA within a specified timeout (e.g., 1 minute), the SSH service
```

Comment:
I believe that the current agentless auth gives 3 minutes to the user to verify operations before giving out, if that can steer the decision for how much time to give in one direction or another. Remember that this happens after the client was shown to be in possession of (regular) Teleport credentials, so it's already a step above being unauthenticated.

Reply:
Silly detail: I'm pretty sure that there's an idle timeout for the bytestream used by the SSH connection, so I think that we should explicitly allow a blank string to be sent by the client that results in a noop and the prompt being sent again, just for the sake of sending and receiving some data.